### PR TITLE
Add if-else function

### PR DESCRIPTION
**Pull request by adam-coogan**
This PR adds an if-else function to keops, addressing discussion in #104. The proposed function takes arguments `f`, `g` and `h`, and returns `g` where `f < 0` and `h` where `f >= 0`. I've tested with the following:
```python
import torch
from pykeops.torch import LazyTensor

ttypes = (torch.FloatTensor, torch.cuda.FloatTensor)

# Test when LazyTensors share underlying data
for ttype in ttypes:
    torch.set_default_tensor_type(ttype)

    # Input
    f = torch.randn([1000, 1])
    f.requires_grad = True
    f_i = LazyTensor(f[:, None, :])

    f_ref = f.detach().clone()
    f_ref.requires_grad = True

    # Values
    first = lambda x: x * x * x
    second = lambda x: x.sqrt()
    val = f_i.ifelse(first(f_i), second(f_i)).sum(dim=0)
    val_refs = torch.zeros_like(f_ref)
    val_refs[f_ref < 0] = first(f_ref[f_ref < 0])
    val_refs[f_ref >= 0] = second(f_ref[f_ref >= 0])
    val_ref = val_refs.sum(dim=0)

    print("Checking values")
    print(f"|val - val_ref|: {(val - val_ref).abs()}\n")
    assert torch.allclose(val, val_ref)

    # Gradients
    val.backward()
    val_ref.backward()
    print("Checking gradients:")
    print(f"max(|grad - grad_ref|): {(f.grad - f_ref.grad).abs().max()}\n")
    assert torch.allclose(f.grad, f_ref.grad)


# Test when other LazyTensors don't share underlying data (maybe unnecessary)
for ttype in ttypes:
    torch.set_default_tensor_type(ttype)

    # Input
    f = torch.randn([1000, 1])
    g = torch.randn_like(f)
    h = torch.randn_like(f).abs()
    f.requires_grad = True
    g.requires_grad = True
    h.requires_grad = True
    f_i = LazyTensor(f[:, None, :])
    g_i = LazyTensor(g[:, None, :])
    h_i = LazyTensor(h[:, None, :])

    f_ref = f.detach().clone()
    g_ref = g.detach().clone()
    h_ref = h.detach().clone()
    f_ref.requires_grad = True
    g_ref.requires_grad = True
    h_ref.requires_grad = True

    # Values
    first = lambda x: x * x * x
    second = lambda x: x.sqrt()
    val = f_i.ifelse(first(g_i), second(h_i)).sum(dim=0)
    val_refs = torch.zeros_like(f_ref)
    val_refs[f_ref < 0] = first(g_ref[f_ref < 0])
    val_refs[f_ref >= 0] = second(h_ref[f_ref >= 0])
    val_ref = val_refs.sum(dim=0)

    print("Checking values")
    print(f"|val - val_ref|: {(val - val_ref).abs()}\n")
    assert torch.allclose(val, val_ref)

    # Gradients
    val.backward()
    val_ref.backward()
    print("Checking gradients:")
    print(f"f: max(|grad|): {f.grad.abs().max()}\n")
    assert torch.allclose(f.grad, torch.zeros_like(f.grad))
    print(f"g: max(|grad - grad_ref|): {(g.grad - g_ref.grad).abs().max()}\n")
    print(f"h: max(|grad - grad_ref|): {(h.grad - h_ref.grad).abs().max()}\n")
    assert torch.allclose(f.grad, torch.zeros_like(f.grad))

```
Please let me know if there's anything I can add/change!

### CODE REVIEW

**Comment by joanglaunes** [comment]
Hello @adam-coogan ,
Thanks for this input ! 
I think it is a very good idea to implement this like you did, with a threshold at zero, and not with the usual assumption "True if non zero", because it allows to get gradients from both expressions.
But now I have one question : why do you take the convention "True for negative, False for positive" ? It seems a bit counter-intuitive to me, but maybe it is standard in some languages or libraries ?

**Comment by adam-coogan** [comment]
Happy to contribute!
> why do you take the convention "True for negative, False for positive" and not the opposite ?
I didn't have a good reason. I'd be happy to switch it if you think that would be more intuitive, just let me know.

**Comment by joanglaunes** [comment]
Ok then yes maybe it will be more intuitive the other way. Can you do the switch please ? Then I'll merge your branch, we will do a new release in a few days hopefully.
Thanks

**Comment by adam-coogan** [comment]
I've made the changes. Here is the new test script:
```python
import torch
import pykeops
from pykeops.torch import LazyTensor


pykeops.clean_pykeops()
ttypes = (torch.FloatTensor, torch.cuda.FloatTensor)

# Test when LazyTensors share underlying data
for ttype in ttypes:
    torch.set_default_tensor_type(ttype)

    # Input
    f = torch.randn([1000, 1])
    f.requires_grad = True
    f_i = LazyTensor(f[:, None, :])

    f_ref = f.detach().clone()
    f_ref.requires_grad = True

    # Values
    first = lambda x: x.sqrt()
    second = lambda x: x * x * x
    # Keops
    val = f_i.ifelse(first(f_i), second(f_i)).sum(dim=0)
    # Torch
    val_refs = torch.zeros_like(f_ref)
    val_refs[f_ref >= 0] = first(f_ref[f_ref >= 0])
    val_refs[f_ref < 0] = second(f_ref[f_ref < 0])
    val_ref = val_refs.sum(dim=0)

    print("Checking values")
    print(f"|val - val_ref|: {(val - val_ref).abs()}\n")
    assert torch.allclose(val, val_ref)

    # Gradients
    val.backward()
    val_ref.backward()
    print("Checking gradients:")
    print(f"max(|grad - grad_ref|): {(f.grad - f_ref.grad).abs().max()}\n")
    assert torch.allclose(f.grad, f_ref.grad)


# Test when other LazyTensors don't share underlying data (maybe unnecessary)
for ttype in ttypes:
    torch.set_default_tensor_type(ttype)

    # Input
    f = torch.randn([1000, 1])
    g = torch.randn_like(f).abs()
    h = torch.randn_like(f)
    f.requires_grad = True
    g.requires_grad = True
    h.requires_grad = True
    f_i = LazyTensor(f[:, None, :])
    g_i = LazyTensor(g[:, None, :])
    h_i = LazyTensor(h[:, None, :])

    f_ref = f.detach().clone()
    g_ref = g.detach().clone()
    h_ref = h.detach().clone()
    f_ref.requires_grad = True
    g_ref.requires_grad = True
    h_ref.requires_grad = True

    # Values
    first = lambda x: x.sqrt()
    second = lambda x: x * x * x
    val = f_i.ifelse(first(g_i), second(h_i)).sum(dim=0)
    val_refs = torch.zeros_like(f_ref)
    val_refs[f_ref >= 0] = first(g_ref[f_ref >= 0])
    val_refs[f_ref < 0] = second(h_ref[f_ref < 0])
    val_ref = val_refs.sum(dim=0)

    print("Checking values")
    print(f"|val - val_ref|: {(val - val_ref).abs()}\n")
    assert torch.allclose(val, val_ref)

    # Gradients
    val.backward()
    val_ref.backward()
    print("Checking gradients:")
    print(f"f: max(|grad|): {f.grad.abs().max()}\n")
    assert torch.allclose(f.grad, torch.zeros_like(f.grad))
    print(f"g: max(|grad - grad_ref|): {(g.grad - g_ref.grad).abs().max()}\n")
    print(f"h: max(|grad - grad_ref|): {(h.grad - h_ref.grad).abs().max()}\n")
    assert torch.allclose(f.grad, torch.zeros_like(f.grad))
```

**Comment by joanglaunes** [comment]
It is merged now, thank you !
